crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+            }
+            Ok(true)
+        })?;
+
+        Ok(ModuleType { defs })
+    }
+
+    fn with_types_scope<T>(&mut self, f: impl FnOnce(&mut Self) -> Result<T>) -> Result<T> {
+        self.types_scopes.push(vec![]);
+        let result = f(self);
+        self.types_scopes.pop();
+        result
+    }
+
+    fn current_type_scope(&self) -> &[Rc<Type>] {
+        self.types_scopes.last().unwrap()
+    }
+
+    fn current_type_scope_mut(&mut self) -> &mut Vec<Rc<Type>> {
+        self.types_scopes.last_mut().unwrap()
+    }
+
+    fn outer_types_scope(&self, count: u32) -> &[Rc<Type>] {
+        &self.types_scopes[self.types_scopes.len() - 1 - usize::try_from(count).unwrap()]
+    }
+
+    fn outer_type(&self, count: u32, i: u32) -> &Rc<Type> {
+        &self.outer_types_scope(count)[usize::try_from(i).unwrap()]
+    }
+
+    fn arbitrary_component_type(&mut self, u: &mut Unstructured<'a>) -> Result<ComponentType> {
+        let mut defs = vec![];
+        let mut imports = HashSet::new();
+        let mut exports = HashSet::new();
+
+        self.with_types_scope(|me| {
+            arbitrary_loop(u, 0, 100, |u| {
+                if !me.current_type_scope().is_empty() && u.arbitrary()? {
+                    // Imports.
+                    let name = crate::unique_string(100, &mut imports, u)?;
+                    let ty = u.int_in_range(0..=me.current_type_scope().len() - 1)?;
+                    let ty = u32::try_from(ty).unwrap();
+                    defs.push(ComponentTypeDef::Import(Import { name, ty }));
+                } else {
+                    // Instance type definitions.
+                    let def = me.arbitrary_instance_type_def(u, &mut exports)?;
+                    defs.push(ComponentTypeDef::Instance(def));
+                }
+                Ok(true)
+            })
+        })?;
+
+        Ok(ComponentType { defs })
+    }
+
+    fn arbitrary_instance_type(&mut self, u: &mut Unstructured<'a>) -> Result<InstanceType> {
+        let mut defs = vec![];
+        let mut exports = HashSet::new();
+
+        self.with_types_scope(|me| {
+            arbitrary_loop(u, 0, 100, |u| {
+                defs.push(me.arbitrary_instance_type_def(u, &mut exports)?);
+                Ok(true)
+            })
+        })?;
+
+        Ok(InstanceType { defs })
+    }
+
+    fn arbitrary_instance_type_def(
+        &mut self,
+        u: &mut Unstructured<'a>,
+        exports: &mut HashSet<String>,
+    ) -> Result<InstanceTypeDef> {
+        match u.int_in_range(0..=2)? {
+            // Export.
+            0 if !self.current_type_scope().is_empty() => Ok(InstanceTypeDef::Export {
+                name: crate::unique_string(100, exports, u)?,
+                ty: u.int_in_range(
+                    0..=u32::try_from(self.current_type_scope().len()).unwrap() - 1,
+                )?,
+            }),
+
+            // Outer type alias.
+            1 if self.types_scopes.iter().any(|scope| !scope.is_empty()) => {
```

Comment:
I think I was thinking that this was equivalent to the `choices`-vec pattern, and so this could avoid the heap allocations, but I think you're right that because the conditions are different it will skew the distributions. I'll switch to using a `choices` vec.